sql/analyzer/stored_procedures.go

```diff
@@ -222,175 +231,199 @@ func validateStoredProcedure(ctx *sql.Context, proc *plan.Procedure) (map[string
 }
 
 // resolveProcedureParams resolves all of the named parameters and declared variables inside of a stored procedure.
-func resolveProcedureParams(ctx *sql.Context, paramNames map[string]struct{}, proc sql.Node) (sql.Node, error) {
-	newProcNode, err := resolveProcedureParamsTransform(ctx, paramNames, proc)
+func resolveProcedureParams(ctx *sql.Context, paramNames map[string]struct{}, proc sql.Node) (sql.Node, transform.TreeIdentity, error) {
+	newProcNode, _, err := resolveProcedureParamsTransform(ctx, paramNames, proc)
 	if err != nil {
-		return nil, err
+		return nil, transform.SameTree, err
 	}
 	// Some nodes do not expose all of their children, so we need to handle them here.
-	newProcNode, err = plan.TransformUp(newProcNode, func(n sql.Node) (sql.Node, error) {
-		switch n := n.(type) {
+	newProc, _, err := transform.Node(newProcNode, func(node sql.Node) (sql.Node, transform.TreeIdentity, error) {
+		switch n := node.(type) {
 		case *plan.InsertInto:
-			newSource, err := resolveProcedureParamsTransform(ctx, paramNames, n.Source)
+			newSource, same, err := resolveProcedureParamsTransform(ctx, paramNames, n.Source)
 			if err != nil {
-				return nil, err
+				return nil, transform.SameTree, err
+			}
+			if same {
+				return n, transform.SameTree, nil
 			}
-			return n.WithSource(newSource), nil
+			return n.WithSource(newSource), transform.NewTree, nil
 		case *plan.Union:
-			newLeft, err := resolveProcedureParamsTransform(ctx, paramNames, n.Left())
+			//TODO unions aren't opaque, is this necessary?
```

Comment:
Unions are opaque, left and right are analyzed in isolation (TransformUp will not descend into children of opaque nodes)